Make `BasisTranslator` rust-native.

raynelfss opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [x] I have added the tests to cover my changes.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
-->

### Summary
The following commits expose a more rust native `BasisTranslator` that operates mostly on the Rust level by taking the following into account:
- Extracting attributes from the `Target` in Rust-space instead of Python.
- Utilizing the rust-native `ParameterExpression` to correctly handle parameters.
- Separate the layers of functionality that are meant to only work with Python.
- Using Rust native errors of the type `BasisTranslatorError` to correclty represent errors without having to go through Python.


### Details and comments
Serves as the prelude for #14874.
~~This PR is currently built on top of #14799 and needs to be rebased once it merges. Rebased now with #14766~~ **REBASED!**

### Pre-requisites:
- [x] #14799
- [x] #14766

raynelfss committed:
TBD

raynelfss committed:
TBD 0

raynelfss committed:
TBD: Moved most of the operations in compose_transform to be rust native.

raynelfss committed:
Fix: Revert any removals related to `ParameterExpression`.
- We cannot get away with not using `ParameterExpression` to create unique parameters. Until a better alternative comes around, we should stick to using it.

raynelfss committed:
Fix: Remove `LazyLock`.
- As of 1.79, `LazyLock` was not yet made stable, these commits remove `LazyLock` in favor of `OnceLock`.

raynelfss committed:
Fix: Remove most pytokens by avoiding building a dict.
- Skipping the dict builds and using `Python::with_gil` avoids using python tokens when not necessary during the `replace_node` method in `run_basis_translator`.
    - The dictionary in question was only really needed when an instance of `ParameterExpression` is present on the gate. So it is optionally built if an instance of Parameter expression is found.
- The last remaining usage of py tokens exists during `compose_transforms` which still builds a parameter vector.

raynelfss committed:
Filter to only parameter expression when building parameter_map.
- In `replace_node` we will first filter out `ParameterExpression` keys in the parameter map. If the collection comes out empty, we don't build the mapping.

raynelfss committed:
Fix: Use rust native parameters.

raynelfss committed:
Fix: Re-add "fixedbitset"

raynelfss committed:
Fix: Convert to `Param::Float` if a real value is obtained

raynelfss commented:
There are some significant speedups:

| Change   | Before [48104583] <stash~10^2>   | After [bb10ab77] <basis-leverage-rust-n>   |   Ratio | Benchmark (Parameter)                                                                                    |
|----------|----------------------------------|--------------------------------------------|---------|----------------------------------------------------------------------------------------------------------|
| -        | 47.7±0.4ms                       | 39.1±0.5ms                                 |    0.82 | passes.MultipleBasisPassBenchmarks.time_basis_translator(14, 1024, ['rx', 'ry', 'rz', 'r', 'rxx', 'id']) |
| -        | 20.4±0.4ms                       | 15.4±0.2ms                                 |    0.75 | passes.MultipleBasisPassBenchmarks.time_basis_translator(5, 1024, ['rx', 'ry', 'rz', 'r', 'rxx', 'id'])  |
| -        | 57.2±2ms                         | 33.8±0.2ms                                 |    0.59 | passes.MultipleBasisPassBenchmarks.time_basis_translator(20, 1024, ['rz', 'x', 'sx', 'cx', 'id'])        |
| -        | 51.8±1ms                         | 25.5±0.5ms                                 |    0.49 | passes.MultipleBasisPassBenchmarks.time_basis_translator(20, 1024, ['u', 'cx', 'id'])                    |
| -        | 18.0±0.2ms                       | 8.47±0.08ms                                |    0.47 | passes.MultipleBasisPassBenchmarks.time_basis_translator(5, 1024, ['rz', 'x', 'sx', 'cx', 'id'])         |
| -        | 37.8±0.2ms                       | 17.4±0.2ms                                 |    0.46 | passes.MultipleBasisPassBenchmarks.time_basis_translator(14, 1024, ['u', 'cx', 'id'])                    |
| -        | 14.7±0.4ms                       | 6.50±0.1ms                                 |    0.44 | passes.MultipleBasisPassBenchmarks.time_basis_translator(5, 1024, ['u', 'cx', 'id'])                     |
| -        | 43.8±1ms                         | 17.7±0.3ms                                 |    0.4  | passes.MultipleBasisPassBenchmarks.time_basis_translator(14, 1024, ['rz', 'x', 'sx', 'cx', 'id'])        |

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.

raynelfss committed:
Fix: Second rebase

raynelfss committed:
Add: Leverage use of rust-native `substitute_node_with_py`.

raynelfss committed:
Fix: Implement native errors for `BasisTranslator`.
- Use `Matrix::Identity` based on the provided matrices.
- Add trait `ToPyErr` that ensures any new error is converted to its respective Python counterpart.
- Remove indexing from `ParameterExpression` creation in `compose_transforms`.
- Remove incorrect coercion into value from `replace_node`.

raynelfss committed:
Add: global operation extraction from Rust.
- Modify `Target::get_non_global_operation_names` to not return an option.

raynelfss committed:
Fix: Remove unused import in python `BasisTranslator`.

Cryoris commented on crates/transpiler/src/passes/basis_translator/errors.rs:
```diff
@@ -0,0 +1,89 @@
+    fn to_py_err(&self) -> PyErr;
+}
+
+impl ToPyError for BasisTranslatorError {
```
Could we just implement
```
impl From<BasisTranslatorError> for PyErr
```
as done for the `ParameterError`, instead of using a new trait? 🙂

Cryoris commented on crates/transpiler/src/passes/basis_translator/errors.rs:
```diff
@@ -0,0 +1,89 @@
+    ]]
+    TargetMissingEquivalence { basis: String, expanded: String },
+    #[error[
+        "Error during creation of DAGCircuit for compose_transforms: {0}"
```
A lot of these errors seem targeting the developers rather than users, but it's the users who will see them. I wonder if we could make these a bit more didactic... e.g. a user won't know what `compose_transforms` is.

raynelfss replied:
Should I just turn it into a generic circuit error? Since I have a lot of them dealing with circuit creation.

Cryoris replied:
Yeah, I would pick just a few more generic categories and assign these ("No equivalence found", "Failed during translation", ..)

Cryoris commented on crates/transpiler/src/passes/basis_translator/compose_transforms.rs:
```diff
@@ -24,129 +36,174 @@ use qiskit_circuit::{
 pub type BasisTransformOut = (SmallVec<[Param; 3]>, DAGCircuit);
 
+static STD_GATE_MAPPING: OnceLock<HashMap<&str, StandardGate>> = OnceLock::new();
+static STD_INST_MAPPING: OnceLock<HashSet<&str>> = OnceLock::new();
```
This is called a "mapping" but it's a hashset -- should we rename this ? 🙂

Cryoris commented on crates/transpiler/src/passes/basis_translator/compose_transforms.rs:
```diff
@@ -24,129 +36,174 @@ use qiskit_circuit::{
+                    ParameterExpression::from_symbol(Symbol::new(
+                        &format!("{gate_name}-{idx}"),
+                        None,
+                        None,
```
We could maybe use the native indexing support here, the symbols have the feature for it
```suggestion
                    ParameterExpression::from_symbol(Symbol::new(
                        gate_name,
                        None,
                        Some(idx),
```

raynelfss replied:
I had it like this before, but was suggested not to use it since there is no `ParameterVector` linked here.

Cryoris replied:
Hmm do we need a linked parameter vector? From the last discussion in the dev meeting we said that the linked vector wouldn't be functionally necessary anywhere, right? 🤔

raynelfss replied:
Added in cbe4e1d868234c50cce87a47723cbac688aea0e7

Cryoris commented on crates/transpiler/src/passes/basis_translator/compose_transforms.rs:
```diff
@@ -24,129 +36,174 @@ use qiskit_circuit::{
+                        None,
+                        None,
+                    ))
+                    .into(),
```
Why is `into()` needed here? Isn't the object already a `ParameterExpression`?

raynelfss replied:
It is but it's not wrapped around an `Arc`.

Cryoris replied:
Ah, sorry -- the `Arc` was added after I went on vacation, I didn't know about it yet 😄

Cryoris commented on crates/transpiler/src/passes/basis_translator/compose_transforms.rs:
```diff
@@ -24,129 +36,174 @@ use qiskit_circuit::{
 
-        let mut dag = DAGCircuit::new()?;
+        let mut dag = DAGCircuit::new()
+            .map_err(|err| BasisTranslatorError::ComposeTransformsCircuitError(err.to_string()))?;
```
Couldn't we just make `DAGCircuit::new` _not_ return a result? Looking at the function it cannot fail and just returns a result out of principle. Then we wouldn't need any error handling here (and we shouldn't need it, since it can't fail 🙂 )

raynelfss replied:
Yeah, but that would fall outside of the scope of this PR, though there shouldn't be any new PRs opening at the moment, we could make an exception for such small change.

Cryoris replied:
It's not a new feature, so I guess a separate PR is fine: https://github.com/Qiskit/qiskit/pull/14923

Cryoris commented on crates/transpiler/src/passes/basis_translator/compose_transforms.rs:
```diff
@@ -24,129 +36,174 @@ use qiskit_circuit::{
-        ))?;
-        let gate_obj: OperationFromPython = gate.extract()?;
+        dag.add_qreg(qubits)
+            .map_err(|err| BasisTranslatorError::ComposeTransformsCircuitError(err.to_string()))?;
```
It seems to me that doing this manual conversion of errors could hide where the error is coming from. How about just having a `BasisTranslatorError::DAGCircuitError(DAGCircuitError)` variant and implement `From<DAGCircuitError> for BasisTranslatorError` -- then you could just call `into()` here and we'd keep track of the original error type.

raynelfss replied:
I like this option much more. After trying it out, these (`DAGCircuitError`, `CircuitError`) error types are Python native and don't implement `Debug` naturally through Rust. For this to work we would need the rust `DAGCircuitError` to be native.

Cryoris commented on crates/transpiler/src/passes/basis_translator/compose_transforms.rs:
```diff
@@ -24,129 +36,174 @@ use qiskit_circuit::{
+                    .call1((&gate_name, gate_num_qubits, placeholder_params.as_ref()))?
+                    .extract()
+            })
+            .map_err(|err| BasisTranslatorError::ComposeTransformsGateError(err.to_string()))?;
```
Based on Matt's discovery, calling `err.to_string()` here requires the GIL, since `err` is a `PyErr` type, right? In this case we should include this in the `with_gil` block.

Cryoris commented on crates/transpiler/src/passes/basis_translator/compose_transforms.rs:
```diff
@@ -24,129 +36,174 @@ use qiskit_circuit::{
     Ok(mapped_instructions)
 }
 
+/// Creates
```
```suggestion
/// Creates the placeholder gate as [PackedOperation].
```

raynelfss replied:
Added in cbe4e1d868234c50cce87a47723cbac688aea0e7

Cryoris commented on crates/transpiler/src/passes/basis_translator/compose_transforms.rs:
```diff
@@ -24,129 +36,174 @@ use qiskit_circuit::{
+        get_standard_gate_names()
+            .iter()
+            .enumerate()
+            .map(|(k, v)| (*v, bytemuck::checked::cast(k as u8)))
```
I'm surprised we need bytemuck here, can't we just do `k as u8`?

raynelfss replied:
Hmm... though `StandardGate` is `repr(u8)` I wasn't able to cast the `u8` to the enum variant.

Cryoris commented on crates/transpiler/src/passes/basis_translator/compose_transforms.rs:
```diff
@@ -24,129 +36,174 @@ use qiskit_circuit::{
+            _ => unreachable!(),
+        };
+        Some(inst.into())
+    } else if name == "unitary" {
```
Is the `unitary` path actually ever executed? I would've thought that we don't have any translations for "unitary" in the equivalence library 🤔 Just having something that allocates an exponential matrix makes me mildly nervous 😛

raynelfss replied:
We don't, that's right. In said case, would you recommend we skip any gates called unitary?

Cryoris replied:
I would probably be explicit about it and put an unreachable stating that we cannot have unitary operations here, since they are not supported in the basis translator

Cryoris commented on crates/transpiler/src/passes/basis_translator/compose_transforms.rs:
```diff
@@ -24,129 +36,174 @@ use qiskit_circuit::{
-                    dag.substitute_node_with_dag(node, &replace_dag, None, None, None)?;
-                }
+                            }
+                            _ => panic!("A non parameter-expression has suck in"),
```
This should be unreachable, right? 😄 
```suggestion
                            _ => unreachable!("A non parameter-expression has snuck in"),
```

raynelfss replied:
Added in cbe4e1d868234c50cce87a47723cbac688aea0e7

Cryoris commented on crates/transpiler/src/passes/basis_translator/mod.rs:
```diff
@@ -437,176 +390,192 @@ fn extract_basis_target_circ(
-    let out_dag = dag.copy_empty_like(VarsMode::Alike)?;
+    let out_dag = dag
+        .copy_empty_like(VarsMode::Alike)
+        .map_err(|err| BasisTranslatorError::ApplyTranslationCircuitError(err.to_string()))?;
```
Same comment with the GIL acquisition here.. we should probably remove `PyErr` from `copy_empty_like` -- looking at the functions that should be straightforward -- and expose a separate `py_copy_empty_like` to Python that has the `PyErr`.

Cryoris replied:
Edit: I wasn't aware that `DAGCircuitError` was a Python error... then this doesn't work.

Cryoris commented on qiskit/transpiler/passes/basis/basis_translator.py:
```diff
@@ -107,12 +105,6 @@ def __init__(self, equivalence_library, target_basis, target=None, min_qubits=0)
         self._non_global_operations = None
         self._qargs_with_non_global_operation = {}
```
Should these attributes now be completely removed?

raynelfss replied:
Removed in cbe4e1d868234c50cce87a47723cbac688aea0e7

Cryoris commented on crates/transpiler/src/passes/basis_translator/mod.rs:
```diff
@@ -47,80 +52,93 @@ use crate::equivalence::EquivalenceLibrary;
+    dag: DAGCircuit,
+    equiv_lib: &mut EquivalenceLibrary,
+    min_qubits: usize,
+    mut target: Option<&mut Target>,
```
Hmm why does the target now need to be mutable, but it wasn't before? It seems like we wouldn't want to change it during basis translation, or do we?

mtreinish replied:
Yeah I have the same question here, we shouldn't ever mutate the target, it defines the constraints of a QPU those are immutable from the transpiler's perspective.

mtreinish replied:
I traced through this, it's a bug in the target. The methods for querying the non-global operations build a cache because this is what the original version in python did. While that was appropriate for Python it's not really expected on a rust getter method and causes mismatched mutability requirements like this. This wasn't caught before because all of this went through python and everything is mutable.

Fundamentally the target api is incorrect right now because of this, especially because in a full transpilation workflow that target is never mutable, the transpiler in C will be given a const pointer to the target and will not be able to mutate it.  If we need caching it should be done in the Python layer.

mtreinish replied:
https://github.com/Qiskit/qiskit/pull/14921 fixes the target api so this doesn't need to be mutable anymore.

mtreinish replied:
In working with this and #14921 locally on my test branch for #14760 there is a knock-on effect of #14921 that it means that a ton of `String` usage in the basis translator after rebasing this PR gets cleaned up to normalize on `&str` and eliminate a ton of allocations and cloning.

mtreinish replied:
Actually that might be too large a change for this PR, it ends up being pretty invasive. I'm pretty sure we can get it avoiding a ton of cloning and allocations by using `&str`. But it's easier to just change this PR with #14921 like:

```diff
diff --git a/crates/transpiler/src/passes/basis_translator/mod.rs b/crates/transpiler/src/passes/basis_translator/mod.rs
index c214081834..2275ca8949 100644
--- a/crates/transpiler/src/passes/basis_translator/mod.rs
+++ b/crates/transpiler/src/passes/basis_translator/mod.rs
@@ -89,7 +89,7 @@ pub fn run_basis_translator(
         let mut qargs_mapping: AhashIndexMap<Qargs, AhashIndexSet<String>> =
             AhashIndexMap::default();
         let global_set: AhashIndexSet<String> =
-            AhashIndexSet::from_iter(target.get_non_global_operation_names(false).iter().cloned());
+            AhashIndexSet::from_iter(target.get_non_global_operation_names(false).into_iter().map(|x| x.to_string()));
         for name in global_set.iter() {
             for qarg in target[name].keys().cloned() {
                 qargs_mapping
```

and when this merges we'll open up an issue to track decreasing the `String` usage in the basis translator.

Cryoris replied:
#14921 is merged now 👍🏻 (remember to also remove the `mut` from `py_run_basis_translator`)

Cryoris commented on crates/transpiler/src/passes/basis_translator/mod.rs:
```diff
@@ -380,25 +344,14 @@ fn extract_basis_target(
 /// this API will be removed with the deprecation of `Pulse`.
 /// TODO: pulse is removed, we can use op.blocks
```
This comment should be removed now, right?

raynelfss replied:
Removed and rewritten in cbe4e1d868234c50cce87a47723cbac688aea0e7

Cryoris commented on crates/transpiler/src/passes/basis_translator/mod.rs:
```diff
@@ -437,176 +390,192 @@ fn extract_basis_target_circ(
+                        let dag_block: DAGCircuit =
+                            circuit_to_dag(block.extract()?, true, None, None)?;
+                        let updated_dag: DAGCircuit;
+                        // TODO: Change this so it makes sense
```
Is this todo still open? It seems yes since the code is unchanged from before, but what does it mean? 😄

raynelfss replied:
Haha, this had to do with how I was implementing the conversion of `BasisTranslatorError` into a `PyError`. It's good now so I'll remove this.

raynelfss replied:
Removed in cbe4e1d868234c50cce87a47723cbac688aea0e7

Cryoris reviewed (commented):
Overall the Python removal looks good to me 👍🏻 my main comments from the first pass are about error handling, namely
1. from our discussions it seems that `PyErr::to_string` requires the GIL, but the GIL is currently not acquired when this method is called -- we should include this call inside the `Python::with_gil` context
2. it would be nice to forward the `DAGCircuitError`s more directly and avoid manual string conversions and wrapping in several custom types. I would suggest to just have a `BasisTranslatorError::DAGCircuitError(DAGCircuitError)` variant for all of these and simply forward the information

Other than that I have a bunch of smaller comments 🙂

raynelfss committed:
Refactor: `ToPyError` to `From<BasisTranslatorError> for PyErr`.

raynelfss committed:
Chore: Rename name set in `compose_transforms`.
- Use acquired gil for `to_string() call of `PyErr`.

mtreinish commented on crates/transpiler/src/passes/basis_translator/mod.rs:
```diff
@@ -47,80 +52,93 @@ use crate::equivalence::EquivalenceLibrary;
+}
+
+pub fn run_basis_translator(
+    dag: DAGCircuit,
```
Why does this take an owned `DAGCircuit`? The basis translator shouldn't need to delete the dagcircuit it receives. I'd expect this to either be `&mut DAGCircuit` if it mutates the dag in place, or `&DAGCircuit` if it only reads the dag and returns a copy.

raynelfss replied:
Fixed in ec026b9

raynelfss committed:
Fix: Take `&mut DAGCircuit` in `BasisTranslator`.

raynelfss committed:
Merge remote-tracking branch 'upstream/main' into basis-leverage-rust-n

mtreinish commented on crates/transpiler/src/passes/basis_translator/mod.rs:
```diff
@@ -47,80 +51,92 @@ use crate::equivalence::EquivalenceLibrary;
+}
+
+pub fn run_basis_translator(
+    dag: &mut DAGCircuit,
```
This doesn't need to be `mut` you're not mutating the dag anywhere. You have L86 calling `as_deref_mut()` but nothing is actually mutating it.

raynelfss replied:
Fixed in cbe4e1d868234c50cce87a47723cbac688aea0e7

raynelfss committed:
Merge remote-tracking branch 'upstream/main' into basis-leverage-rust-n

raynelfss committed:
Fix: Address rest of review comments

Co-authored-by: Julien Gacon <gaconju@gmail.com>

Cryoris commented on crates/transpiler/src/passes/basis_translator/compose_transforms.rs:
```diff
@@ -24,119 +35,173 @@ use qiskit_circuit::{
-            .get_bound(py)
-            .call1((&gate_name, num_params))?
-            .extract()?;
+        // The last usage of Python left is the parameter vector here.
```
Is this comment still up to date, or was this referring to when it was previously using `PARAMETER_VECTOR` here?

raynelfss replied:
Removed in a210bcfb9eeef47be38bc7dab418ed2f7378ee67

Cryoris commented on crates/transpiler/src/passes/basis_translator/compose_transforms.rs:
```diff
@@ -24,119 +35,173 @@ use qiskit_circuit::{
             #[cfg(feature = "cache_pygates")]
-            Some(gate.into()),
-        )?;
+            None,
```
This reminds me that we should probably do a small benchmark to ensure we didn't accidentally slow anything down by removing caches (I don't think we should be slower, but better double check)

raynelfss replied:
ASV doesn't seem to detect any dramatic slowdown:

Basis Translator benchmarks:

| Change   | Before [25c8a931] <fix-c-opscount-free~2>   | After [a210bcfb] <basis-leverage-rust-n>   |   Ratio | Benchmark (Parameter)                                                                                    |
|----------|---------------------------------------------|--------------------------------------------|---------|----------------------------------------------------------------------------------------------------------|
| -        | 55.7±1ms                                    | 34.1±0.6ms                                 |    0.61 | passes.MultipleBasisPassBenchmarks.time_basis_translator(20, 1024, ['rx', 'ry', 'rz', 'r', 'rxx', 'id']) |
| -        | 40.8±0.5ms                                  | 24.6±0.8ms                                 |    0.6  | passes.MultipleBasisPassBenchmarks.time_basis_translator(14, 1024, ['rx', 'ry', 'rz', 'r', 'rxx', 'id']) |
| -        | 46.9±0.9ms                                  | 24.9±2ms                                   |    0.53 | passes.MultipleBasisPassBenchmarks.time_basis_translator(20, 1024, ['rz', 'x', 'sx', 'cx', 'id'])        |
| -        | 17.5±0.4ms                                  | 9.27±0.4ms                                 |    0.53 | passes.MultipleBasisPassBenchmarks.time_basis_translator(5, 1024, ['rx', 'ry', 'rz', 'r', 'rxx', 'id'])  |
| -        | 35.5±2ms                                    | 18.5±1ms                                   |    0.52 | passes.MultipleBasisPassBenchmarks.time_basis_translator(14, 1024, ['rz', 'x', 'sx', 'cx', 'id'])        |
| -        | 41.2±1ms                                    | 21.0±1ms                                   |    0.51 | passes.MultipleBasisPassBenchmarks.time_basis_translator(20, 1024, ['u', 'cx', 'id'])                    |
| -        | 30.9±0.9ms                                  | 15.1±0.4ms                                 |    0.49 | passes.MultipleBasisPassBenchmarks.time_basis_translator(14, 1024, ['u', 'cx', 'id'])                    |
| -        | 13.8±0.4ms                                  | 6.62±0.2ms                                 |    0.48 | passes.MultipleBasisPassBenchmarks.time_basis_translator(5, 1024, ['rz', 'x', 'sx', 'cx', 'id'])         |
| -        | 12.5±0.3ms                                  | 4.99±0.3ms                                 |    0.4  | passes.MultipleBasisPassBenchmarks.time_basis_translator(5, 1024, ['u', 'cx', 'id'])                     |

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.

---

Utility Scale
| Change   | Before [25c8a931] <fix-c-opscount-free~2>   | After [a210bcfb] <basis-leverage-rust-n>   | Ratio   | Benchmark (Parameter)                                                         |
|----------|---------------------------------------------|--------------------------------------------|---------|-------------------------------------------------------------------------------|
|          | 0                                           | 0                                          | n/a     | utility_scale.UtilityScaleBenchmarks.track_bvlike_depth('cx')                 |
|          | 0                                           | 0                                          | n/a     | utility_scale.UtilityScaleBenchmarks.track_bvlike_depth('cz')                 |
|          | 0                                           | 0                                          | n/a     | utility_scale.UtilityScaleBenchmarks.track_bvlike_depth('ecr')                |
|          | 4.90±0.04ms                                 | 5.35±0.04ms                                | 1.09    | utility_scale.UtilityScaleBenchmarks.time_bvlike('ecr')                       |
|          | 4.93±0.02ms                                 | 5.34±0.07ms                                | 1.08    | utility_scale.UtilityScaleBenchmarks.time_bvlike('cx')                        |
|          | 4.95±0.06ms                                 | 5.35±0.09ms                                | 1.08    | utility_scale.UtilityScaleBenchmarks.time_bvlike('cz')                        |
|          | 66.8±0.6ms                                  | 67.9±0.5ms                                 | 1.02    | utility_scale.UtilityScaleBenchmarks.time_square_heisenberg('cx')             |
|          | 75.5±0.4ms                                  | 76.9±0.7ms                                 | 1.02    | utility_scale.UtilityScaleBenchmarks.time_square_heisenberg('cz')             |
|          | 74.9±0.2ms                                  | 76.3±2ms                                   | 1.02    | utility_scale.UtilityScaleBenchmarks.time_square_heisenberg('ecr')            |
|          | 44.9±0.3ms                                  | 45.5±0.5ms                                 | 1.01    | utility_scale.UtilityScaleBenchmarks.time_bv_100('cx')                        |
|          | 47.1±0.7ms                                  | 46.9±0.4ms                                 | 1.00    | utility_scale.UtilityScaleBenchmarks.time_bv_100('cz')                        |
|          | 47.4±0.6ms                                  | 47.5±0.5ms                                 | 1.00    | utility_scale.UtilityScaleBenchmarks.time_bv_100('ecr')                       |
|          | 72.9±0.4ms                                  | 72.6±0.3ms                                 | 1.00    | utility_scale.UtilityScaleBenchmarks.time_parse_qft_n100('cx')                |
|          | 73.0±0.1ms                                  | 72.9±0.7ms                                 | 1.00    | utility_scale.UtilityScaleBenchmarks.time_parse_qft_n100('ecr')               |
|          | 23.2±0.09ms                                 | 23.3±0.1ms                                 | 1.00    | utility_scale.UtilityScaleBenchmarks.time_parse_square_heisenberg_n100('cz')  |
|          | 254±1ms                                     | 254±2ms                                    | 1.00    | utility_scale.UtilityScaleBenchmarks.time_qaoa('cz')                          |
|          | 357±1ms                                     | 357±3ms                                    | 1.00    | utility_scale.UtilityScaleBenchmarks.time_qft('cx')                           |
|          | 427±4ms                                     | 429±4ms                                    | 1.00    | utility_scale.UtilityScaleBenchmarks.time_qft('cz')                           |
|          | 437±3ms                                     | 436±3ms                                    | 1.00    | utility_scale.UtilityScaleBenchmarks.time_qft('ecr')                          |
|          | 513±4ms                                     | 512±3ms                                    | 1.00    | utility_scale.UtilityScaleBenchmarks.time_qv('cx')                            |
|          | 565±3ms                                     | 562±4ms                                    | 1.00    | utility_scale.UtilityScaleBenchmarks.time_qv('cz')                            |
|          | 400                                         | 400                                        | 1.00    | utility_scale.UtilityScaleBenchmarks.track_bv_100_depth('cx')                 |
|          | 400                                         | 400                                        | 1.00    | utility_scale.UtilityScaleBenchmarks.track_bv_100_depth('cz')                 |
|          | 400                                         | 400                                        | 1.00    | utility_scale.UtilityScaleBenchmarks.track_bv_100_depth('ecr')                |
|          | 300                                         | 300                                        | 1.00    | utility_scale.UtilityScaleBenchmarks.track_circSU2_depth('cx')                |
|          | 300                                         | 300                                        | 1.00    | utility_scale.UtilityScaleBenchmarks.track_circSU2_depth('cz')                |
|          | 300                                         | 300                                        | 1.00    | utility_scale.UtilityScaleBenchmarks.track_circSU2_depth('ecr')               |
|          | 1590                                        | 1590                                       | 1.00    | utility_scale.UtilityScaleBenchmarks.track_qaoa_depth('cx')                   |
|          | 1603                                        | 1603                                       | 1.00    | utility_scale.UtilityScaleBenchmarks.track_qaoa_depth('cz')                   |
|          | 1603                                        | 1603                                       | 1.00    | utility_scale.UtilityScaleBenchmarks.track_qaoa_depth('ecr')                  |
|          | 2692                                        | 2692                                       | 1.00    | utility_scale.UtilityScaleBenchmarks.track_qft_depth('cx')                    |
|          | 2744                                        | 2744                                       | 1.00    | utility_scale.UtilityScaleBenchmarks.track_qft_depth('cz')                    |
|          | 2744                                        | 2744                                       | 1.00    | utility_scale.UtilityScaleBenchmarks.track_qft_depth('ecr')                   |
|          | 2571                                        | 2571                                       | 1.00    | utility_scale.UtilityScaleBenchmarks.track_qv_depth('cx')                     |
|          | 2571                                        | 2571                                       | 1.00    | utility_scale.UtilityScaleBenchmarks.track_qv_depth('cz')                     |
|          | 2571                                        | 2571                                       | 1.00    | utility_scale.UtilityScaleBenchmarks.track_qv_depth('ecr')                    |
|          | 480                                         | 480                                        | 1.00    | utility_scale.UtilityScaleBenchmarks.track_square_heisenberg_depth('cx')      |
|          | 480                                         | 480                                        | 1.00    | utility_scale.UtilityScaleBenchmarks.track_square_heisenberg_depth('cz')      |
|          | 480                                         | 480                                        | 1.00    | utility_scale.UtilityScaleBenchmarks.track_square_heisenberg_depth('ecr')     |
|          | 502±6ms                                     | 499±10ms                                   | 0.99    | utility_scale.UtilityScaleBenchmarks.time_circSU2('cz')                       |
|          | 6.73±0.04ms                                 | 6.63±0.02ms                                | 0.99    | utility_scale.UtilityScaleBenchmarks.time_parse_qaoa_n100('cx')               |
|          | 6.71±0.06ms                                 | 6.63±0.02ms                                | 0.99    | utility_scale.UtilityScaleBenchmarks.time_parse_qaoa_n100('ecr')              |
|          | 72.9±0.2ms                                  | 72.2±0.2ms                                 | 0.99    | utility_scale.UtilityScaleBenchmarks.time_parse_qft_n100('cz')                |
|          | 23.5±0.2ms                                  | 23.3±0.2ms                                 | 0.99    | utility_scale.UtilityScaleBenchmarks.time_parse_square_heisenberg_n100('cx')  |
|          | 23.5±0.2ms                                  | 23.1±0.1ms                                 | 0.99    | utility_scale.UtilityScaleBenchmarks.time_parse_square_heisenberg_n100('ecr') |
|          | 557±5ms                                     | 554±5ms                                    | 0.99    | utility_scale.UtilityScaleBenchmarks.time_qv('ecr')                           |
|          | 6.72±0.03ms                                 | 6.59±0.03ms                                | 0.98    | utility_scale.UtilityScaleBenchmarks.time_parse_qaoa_n100('cz')               |
|          | 208±4ms                                     | 205±2ms                                    | 0.98    | utility_scale.UtilityScaleBenchmarks.time_qaoa('cx')                          |
|          | 242±1ms                                     | 237±2ms                                    | 0.98    | utility_scale.UtilityScaleBenchmarks.time_qaoa('ecr')                         |
|          | 502±10ms                                    | 481±8ms                                    | 0.96    | utility_scale.UtilityScaleBenchmarks.time_circSU2('ecr')                      |
|          | 499±6ms                                     | 476±6ms                                    | 0.95    | utility_scale.UtilityScaleBenchmarks.time_circSU2('cx')                       |

Cryoris replied:
Great, thanks for checking!

Cryoris commented on crates/transpiler/src/passes/basis_translator/compose_transforms.rs:
```diff
@@ -24,119 +35,173 @@ use qiskit_circuit::{
-                    dag.substitute_node_with_dag(node, &replace_dag, None, None, None)?;
-                }
+                            }
+                            _ => unreachable!("A non parameter-expression has suck in"),
```
```suggestion
                            _ => unreachable!("A non parameter-expression has snuck in"),
```

raynelfss replied:
Corrected in a210bcfb9eeef47be38bc7dab418ed2f7378ee67 (Sorry it took so long :sob:)

Cryoris commented on crates/transpiler/src/passes/basis_translator/mod.rs:
```diff
@@ -47,80 +51,92 @@ use crate::equivalence::EquivalenceLibrary;
-#[pyfunction(name = "base_run", signature = (dag, equiv_lib, qargs_with_non_global_operation, min_qubits, target_basis=None, target=None, non_global_operations=None))]
+#[pyfunction(name = "base_run", signature = (dag, equiv_lib, min_qubits, target=None, target_basis=None))]
+fn py_run_basis_translator(
+    dag: &mut DAGCircuit,
```
This also doesn't need to be mutable, since it's passed as `&DAGCircuit` into `run_basis_translator` 🙂

raynelfss replied:
Changed in a210bcfb9eeef47be38bc7dab418ed2f7378ee67

Cryoris commented on crates/transpiler/src/passes/basis_translator/mod.rs:
```diff
@@ -437,176 +384,191 @@ fn extract_basis_target_circ(
+                    for block in blocks.try_iter()? {
+                        let block = block?;
+                        let dag_block: DAGCircuit =
+                            circuit_to_dag(block.extract()?, true, None, None)?;
```
Even though we have the GIL, it'd be cleaner to avoid using it, right? Can we do this as:
```suggestion
                    for block in node_obj.op.block() {
                        let dag_block: DAGCircuit = circuit_to_dag(&block, true, None, None)?;
```

raynelfss replied:
For this we specifically need to end up with `QuantumCircuit` instances. And if we do this here we will end up having to convert the blocks to `QuantumCircuit` instances. This is due to `ControlFlowOp` taking in

raynelfss replied:
I updated it to use your methodology in a210bcfb9eeef47be38bc7dab418ed2f7378ee67, but this will not ever look good until #14568 merges.

Cryoris commented on crates/transpiler/src/passes/basis_translator/mod.rs:
```diff
@@ -437,176 +384,191 @@ fn extract_basis_target_circ(
+                            qargs_with_non_global_operation,
+                        ).map_err(PyErr::from)?;
+                        let flow_circ_block = if is_updated {
+                            DAG_TO_CIRCUIT
```
Could we use Rust's `qiskit_circuit::converters::dag_to_circuit` here?

raynelfss replied:
Added in c7d2a7709cc9a7fe59218ca12dd69fd983af857b

Cryoris commented on crates/transpiler/src/passes/basis_translator/mod.rs:
```diff
@@ -754,47 +792,89 @@ fn replace_node(
-                            new_phase.call_method1(intern!(py, "assign"), key_val.downcast()?)?;
+            Param::ParameterExpression(old_phase) => {
+                let new_phase: Param = {
+                    let mut bind_dict = HashMap::new();
```
We could use `with_capacity` here if we added a method to get the number of symbols to `ParameterExpression`. That should be fairly easy, we'd just need
```rs
impl ParameterExpression { 
  /// Get the number of [Symbol]s in the expression.
  pub fn num_symbols(&self) -> usize {
    self.name_map.len()
  }
}
```

raynelfss replied:
Added in c7d2a7709cc9a7fe59218ca12dd69fd983af857b

Cryoris commented on crates/transpiler/src/passes/basis_translator/mod.rs:
```diff
@@ -689,57 +680,104 @@ fn replace_node(
+                                            .into_iter()
+                                            .collect();
+                                    new_value = new_value
+                                        .subs(&map, false)
```
If we're sure that the `map` is valid, we could change this to
```suggestion
                                        .subs(&map, true)
```
which would skip some symbol checks and be faster (same in other places for `bind`)

raynelfss replied:
Added in c7d2a7709cc9a7fe59218ca12dd69fd983af857b

Cryoris commented on crates/transpiler/src/passes/basis_translator/mod.rs:
```diff
@@ -689,57 +680,104 @@ fn replace_node(
-                                    intern!(py, "assign"),
-                                    items.downcast::<PyTuple>()?,
-                                )?;
+                        let new_value: Param = {
```
If I understand correctly, this whole bit assigns the `Param`s from `parameter_map` to the `param` object in the `inner_node`, right? 

If yes, this looks a bit convoluted right now... couldn't we just 
1. create two hashmaps, one for calling `subs` and one for `bind`
2. iterate over matching elements in `parameter_map` and push onto the correct hashmap depending on the value 
3. call subs and bind with the hashmaps

And maybe we should wrap this in a function since it is used again below, right?

In a follow-up we should support `ParameterExpression::assign(symbol: &Symbol, param: Param)` since this is also used in the circuit assignment.

raynelfss replied:
I like this idea, specially since it allows us to re-use the custom assignment workflow. I will try and adapt this.

raynelfss replied:
Added it in 9d3aad6b52cd69836af36e12d7f613198b30c4f2

raynelfss committed:
Merge remote-tracking branch 'upstream/main' into basis-leverage-rust-n

raynelfss committed:
Refactor: Use `&str` over `String` when possible.

raynelfss committed:
Chore: Round of review comments.

raynelfss committed:
Fix: More review comments
- Use rust native dag_to_circuit method and convert to `QuantumCircuit` for control flow procedures.
-  Add `num_symbols` to `ParameterExpression`.
- Pass `allow_unknwon_parameters` as true for `bind` and `subs` in `BasisTranslator`.

Co-authored-by: Julien Gacon <gaconju@gmail.com>

raynelfss committed:
Fix: Split out parameter assignment from `replace_node`

raynelfss committed:
Fix: More efficient parameter binding

Co-authored-by: Julien Gacon <gaconju@gmail.com>

raynelfss committed:
Merge remote-tracking branch 'upstream/main' into basis-leverage-rust-n

Cryoris commented on crates/transpiler/src/passes/basis_translator/mod.rs:
```diff
@@ -796,92 +710,96 @@ fn replace_node(
+    for key in param_obj.iter_symbols() {
+        match &parameter_map[key].clone() {
+            Param::ParameterExpression(val) => {
+                subs_map.insert(key.clone(), val.as_ref().clone());
```
Oh we should update `subs` to also take `&Symbol` as  key in a separate PR

Cryoris commented on crates/transpiler/src/passes/basis_translator/mod.rs:
```diff
@@ -796,92 +710,96 @@ fn replace_node(
+                Ok(parsed) => Ok(Param::ParameterExpression(
+                    ParameterExpression::from_symbol_expr(SymbolExpr::Value(parsed)).into(),
+                )),
+                Err(_) => Ok(Param::ParameterExpression(new_value.into())),
```
This cannot be reached since you're checking above that there are no symbols in the map -- I would just rewrite this without the `iter_symbols.next.is_none` bit and just match on `try_to_value` directly instead 🙂

raynelfss replied:
That would need me to make the call strict too I believe.

raynelfss replied:
Added in 7cb647cdd4e84ce1b767112262edb910208f5ffb

Cryoris commented on crates/transpiler/src/passes/basis_translator/mod.rs:
```diff
@@ -796,92 +710,96 @@ fn replace_node(
-                    BasisTranslatorError::ReplaceNodeCircuitError(err.to_string())
-                })?;
-            }
+        param_assignment_global_phase(target_dag.global_phase(), &parameter_map, dag)?;
```
Could we simplify this and remove the additional function if we'd 
1. call `param_assignment_expr` to get the Param
2. match to assert there's no complex value
3. call `dag.add_global_phase`
?

raynelfss replied:
We can just add an `allow_complex` argument in `param_assignment_expr` instead.

raynelfss replied:
Added in 7cb647cdd4e84ce1b767112262edb910208f5ffb

Cryoris commented on crates/transpiler/src/passes/basis_translator/mod.rs:
```diff
@@ -796,92 +710,96 @@ fn replace_node(
+}
 
-            _ => {}
+fn param_assignment_expr(
```
Could we make this a method on `Param`? We could also do that in a follow-up if you prefer.

Cryoris commented on crates/transpiler/src/passes/basis_translator/mod.rs:
```diff
@@ -223,87 +215,59 @@ pub fn run_basis_translator(
-        Ok(())
     }
 
     // Recurse for QuantumCircuit
```
```suggestion
    // Recurse for CircuitData
```
.... or just delete, the function name seems clear enough 😉

raynelfss replied:
Fixed in 7cb647cdd4e84ce1b767112262edb910208f5ffb

Cryoris commented on crates/transpiler/src/passes/basis_translator/mod.rs:
```diff
@@ -437,176 +377,185 @@ fn extract_basis_target_circ(
-    let out_dag = dag.copy_empty_like(VarsMode::Alike)?;
+    let out_dag = dag
+        .copy_empty_like(VarsMode::Alike)
+        .map_err(|err| BasisTranslatorError::ApplyTranslationCircuitError(err.to_string()))?;
```
What was the conclusion on calling `err.to_string()` now? Should we just map this to a new error without calling `to_string`? E.g.
```suggestion
        .map_err(|_| BasisTranslatorError::TranslationFailed)?;
```

raynelfss replied:
I just made generic `Circuit` and `DAGCircuit` errors that may be replaced in the future when we implement rust native errors for the rest of the API. Once that's done we can change the errors to use `#[from] CircuitError` when that is done. I'll add `TODO` notes.

raynelfss replied:
Added in baabe1c1fc8d03bfbf2be07418f4799d44bb800e

raynelfss committed:
Remove: Path for unitaries.
As per @Cryoris previous reviews, there is no path for a `UnitaryGate` to be represented as a key in the `EquivalenceLibrary`. Thus we can remove the unitary path from the `name_to_packed_operation` method.

raynelfss committed:
Fix: Remove most string conversions from `BasisTranslatorError`.

raynelfss committed:
Fix: Address first wave of review items
- Fix comment in `recurse_circuit`.
- Remove `param_assignment_global_phase` and use `param_assignment_expr.
- `param_assignment_expr` now accepts a `ParameterExpression` instance.
- Add `allow_complex` flag in `param_assignment_expr` to prevent complex values from getting assigned.

raynelfss committed:
Merge branch 'main' into basis-leverage-rust-n

Cryoris commented on crates/transpiler/src/passes/basis_translator/mod.rs:
```diff
@@ -620,19 +577,25 @@ fn replace_node(
+            let new_op: PackedOperation = match inner_node.op.view() {
+                OperationRef::Gate(gate) => {
+                    Python::with_gil(|py| gate.py_copy(py).map(|op| op.into()))
+                        .map_err(|err| BasisTranslatorError::BasisGateError(err.to_string()))?
```
Here you don't have the GIL anymore when calling `err.to_string()`, right? Then we also need to replace this with an hardcoded string error, right? Same in a bunch of other places 🙂

raynelfss replied:
Oh, since we're still going to use the GIL anyway. We can just move the message to be obtained inside of the GIL.

Cryoris commented on crates/transpiler/src/passes/basis_translator/errors.rs:
```diff
@@ -0,0 +1,85 @@
+        "{0}"
+    ]]
+    BasisCircuitError(String),
+    // TODO: Use rust native GateError
```
Do we have a `GateError`? I couldn't find one

raynelfss replied:
We actually don't have one. Most gate errors seem to be regular errors. What should I replace it with? Should I remove it?

Cryoris replied:
Hm maybe just `.expect(...)` these errors, since they are just propagating potential errors from `py_copy` which shouldn't fail?

raynelfss replied:
Added in e2f04f55e9fc9ecab0b459aeb526c719d246f90b

Cryoris commented on crates/transpiler/src/passes/basis_translator/errors.rs:
```diff
@@ -0,0 +1,85 @@
+    #[error[
+        "Error mapping of parameters for replace_node: {0}"
+    ]]
+    ReplaceNodeParameterError(String),
```
This one doesn't seem very intuitive to understand as user, could we just fail more generically here? Maybe grouped with the above and just say that parameter assignment during translation failed or sth?

raynelfss replied:
Yeah I can put this and the global phase error together. My only worry is that we lose specificity with the errors by generalizing so much. Ideally we'd want to use `this-error` to have a more holistic understanding of the error in question by having these variants. Since the Rust API is mostly developer facing, I think leaving these separate is okay.

That said, we can rename the ReplaceNodeParameterError to just re-use the `ParameterError` returned from the source and make it look more like this:
```rs
/// A collection of the Errrors possible in the [run_basis_translator] function
#[derive(Debug, Error)]
pub enum BasisTranslatorError {
    //...
    #[error(transparent)]
    BasisParameterError(#[from] ParameterError),   
    //...
}

raynelfss replied:
Added in 8be5008f162d15b9cf577df46143e4a3abcd4d14

raynelfss committed:
Refactor: Use `ParameterError` as a base for `BasisTranslatorError`.

raynelfss committed:
FIx: Remove `BasisGateError`, use `.expect()` for the rest.

Cryoris commented on crates/transpiler/src/passes/basis_translator/compose_transforms.rs:
```diff
@@ -24,119 +31,177 @@ use qiskit_circuit::{
+                        .call1((&gate_name, gate_num_qubits, placeholder_params.as_ref()))?
+                        .extract()
+                };
+                gate()
```
Is there a reason for this lambda function? It would seem simpler to just create the gate directly 😄

raynelfss replied:
No reason :/

Cryoris commented on crates/transpiler/src/passes/basis_translator/compose_transforms.rs:
```diff
@@ -24,119 +31,177 @@ use qiskit_circuit::{
+                };
+                gate()
+            })
+            .unwrap_or_else(|_| panic!("Error creating custom gate for entry {}", gate_name));
```
```suggestion
            .expect(format!("Error creating custom gate for entry {}", gate_name)));
```
seems more idiomatic 🙂

raynelfss replied:
This was done by clippy because `.expect` requires an `&str`. Since `format!` returns a `String` clippy automatically corrects it.

Cryoris replied:
Lol ok -- I learned something new, thanks!

Cryoris commented on crates/transpiler/src/passes/basis_translator/compose_transforms.rs:
```diff
@@ -24,119 +31,177 @@ use qiskit_circuit::{
-                    })
-                    .collect::<Vec<_>>();
-                for (node, params) in nodes_to_replace {
-                    let param_mapping: IndexMap<ParameterUuid, Param, ahash::RandomState> =
-                        equiv_params
+    for ((gate_name, gate_num_qubits), (equiv_params, equiv)) in basis_transforms {
```
I'm a bit confused by this change in the loop: previously, we iterated over `basis_transforms` in the same loop as the `source_basis` loop, but now they are sequential. Is this change intentional?

raynelfss replied:
Yeah, I had to go back to the original `BasisTranslator` code and realized this loop wasn't supposed to run inside of the `source_basis` loop. This was a bug that snuck in from the previous iteration.

Check this snippet from the original: https://github.com/Qiskit/qiskit/blob/0bc84296f92d8eb9d51f52a4ca117a87804fed31/qiskit/transpiler/passes/basis/basis_translator.py#L626-L651

Cryoris replied:
Ok nicely caught then 🙂

Cryoris commented on crates/transpiler/src/passes/basis_translator/compose_transforms.rs:
```diff
@@ -24,119 +31,177 @@ use qiskit_circuit::{
-            .get_bound(py)
-            .call1((&gate_name, num_params))?
-            .extract()?;
+        let mut placeholder_params: SmallVec<[Param; 3]> = (0..num_params as u32)
```
Do these need to be `mut`? I don't think we change them, do we?

raynelfss replied:
We change them when a python gate is made and we re-extract the parameters from there. I can have it removed though.

Cryoris replied:
Ah right, no then they are indeed `mut`

Cryoris commented on crates/transpiler/src/passes/basis_translator/compose_transforms.rs:
```diff
@@ -24,119 +31,177 @@ use qiskit_circuit::{
+    if let Some(operation) = std_gate_mapping.get(name) {
+        Some((*operation).into())
+    } else if std_instruction_mapping.contains(name) {
+        let inst = match name {
+            "barrier" => StandardInstruction::Barrier(num_qubits),
+            "delay" => StandardInstruction::Delay(qiskit_circuit::operations::DelayUnit::DT),
+            "measure" => StandardInstruction::Measure,
+            "reset" => StandardInstruction::Reset,
```
I don't think we contain translations for instructions, do we? I think we can also replace this whole block with an `unreachable` statement like for the unitary

raynelfss replied:
Sure! Couldn't remove this unfortunately. The specified target basis doesn't filter out the `StandardInstructions`.

Cryoris commented on crates/transpiler/src/passes/basis_translator/errors.rs:
```diff
@@ -0,0 +1,82 @@
+    #[error[
+    "{0}"
+    ]]
+    // TODO: Use rust native DAGCircuitError
```
```suggestion
    // TODO: Use rust native DAGCircuitError
    #[error[
    "{0}"
    ]]
```

raynelfss replied:
Fixed in 523a30b09cc25437be01f7e4850cd74e63771dd8

Cryoris commented on crates/transpiler/src/passes/basis_translator/mod.rs:
```diff
@@ -47,99 +50,105 @@ use crate::equivalence::EquivalenceLibrary;
+    equiv_lib: &mut EquivalenceLibrary,
+    min_qubits: usize,
+    target: Option<&Target>,
+    target_basis: Option<HashSet<String>>,
```
We should take an `Option<&HashSet<&str>>` in the Rust interface I think, and create the correct type in the python wrapper above 🤔

mtreinish replied:
In general I agree we don't need or want an owned type here, but in practice we should never be setting this from Rust, it should always be `None`. In rust space we should always be using a `Target` and never loose constraints, this is strictly a backwards compatibility thing for Python.

It would be better if we just removed this argument from the rust code and did `Target.from_configuration(basis_gates)` in the Python code for the pass instead and pass the generated target to rust. That would also simplify the rust code so we only have a target to worry about.

raynelfss replied:
Added in 523a30b09cc25437be01f7e4850cd74e63771dd8

raynelfss committed:
Fix: Final review comments?
- Remove hashset from `compose_transforms` use array too.
- Fix comment in error.
- Build `HashSet<&str>` from python counterpart and use as main `target_basis` argument in `run_basis_translator`.

raynelfss committed:
Fix: Re-add removed parameter extraction

Cryoris reviewed (approved):
Thanks for all the work, this looks in a good shape to me now to unblock the C transpiler function. There's some follow-ups which we can do but with the beta release in mind I would be happy to merge it 🙂 

The main follow ups are: 
* https://github.com/Qiskit/qiskit/pull/14659#discussion_r2298495831 (this we could still do here if you want, but I'm fine doing a follow up to unblock now)
* the Rust-native errors
* putting parameter assignments as methods on `Param` and `ParameterExpression` instead of standalones
